src/caffe/layers/base_data_layer.cu

```diff
@@ -13,6 +13,9 @@ void BasePrefetchingDataLayer<Dtype>::Forward_gpu(
   // Copy the data
   caffe_copy(batch->data_.count(), batch->data_.gpu_data(),
       top[0]->mutable_gpu_data());
+  // Ensure the copy is synchronous wrt the host, so that the next batch isn't
+  // copied in meanwhile.
+  CUDA_CHECK(cudaStreamSynchronize(cudaStreamLegacy));
```

Comment:
Thanks @longjon for catching this and the explanation.

However, I am still a little confused why calling `cudaStreamSynchronize(cudaStreamLegacy)` here, rather than several lines below after `batch->label_` is also copied. I think this synchronization should occur right before `prefetch_free_.push(batch)` (correct me if I'm wrong)